### PR TITLE
Second round of perf changes

### PR DESCRIPTION
**Pull request by chrisforbes**
Chops off another 25s / 15% or so (so far) off my internal workload.

This is very WIP -- there are a number of tests that assume a pattern is a std::deque (and some of them just verify that a deque works...; also various TODO dropped in the code where I can do better.)

### CODE REVIEW

File: source/operand.cpp [hunk 1]
```diff
@@ -262,24 +267,23 @@ bool spvExpandOperandSequenceOnce(spv_operand_type_t type,
                                   spv_operand_pattern_t* pattern) {
   switch (type) {
     case SPV_OPERAND_TYPE_VARIABLE_ID:
-      pattern->insert(pattern->begin(), {SPV_OPERAND_TYPE_OPTIONAL_ID, type});
+      pattern->insert(pattern->end(), {type, SPV_OPERAND_TYPE_OPTIONAL_ID});
```
Comment (atgoo) [inline source/operand.cpp]: Is this not just emplace_back?

File: source/operand.cpp [hunk 2]
```diff
@@ -292,25 +296,29 @@ spv_operand_type_t spvTakeFirstMatchableOperand(
...
-    alternatePattern.push_back(SPV_OPERAND_TYPE_OPTIONAL_CIV);
+  // TODO: make this not terrible again: scan for RESULT_ID first,
+  // then decide what to do. (chrisforbes) not really bothered by
+  // this being a bit slow-- is a weird case in assembler path.
+
+  for (auto it = pattern.crbegin(); it != pattern.crend(); it++) {
+      if (*it == SPV_OPERAND_TYPE_RESULT_ID) {
+          alternatePattern.insert(alternatePattern.begin(), *it);
+          alternatePattern.insert(alternatePattern.begin(), SPV_OPERAND_TYPE_OPTIONAL_CIV);
+          return alternatePattern;
```
Comment (atgoo) [inline source/operand.cpp]: Consider
```
spv_operand_pattern_t alternatePattern(index + 2, SPV_OPERAND_TYPE_OPTIONAL_CIV);
alternatePattern[index + 1] = SPV_OPERAND_TYPE_RESULT_ID;
return alternatePattern;
```

Comment (atgoo) [inline source/operand.cpp]: I don't know what this function is doing but it looks like we want to return
{o, x, o, ..., o}
if result_id is found, or
{o}
if not.

File: source/operand.cpp (hunk 1 above)
Comment (chrisforbes) [inline source/operand.cpp]: It can be, yes.

File: source/operand.cpp (hunk 2 above)
Comment (chrisforbes) [inline source/operand.cpp]: Will do something along those lines. I'd not really cared very much about making this function good -- doesn't appear in my profile at all -- but not difficult to do this right.

**Comment by dneto0** [comment]
I haven't looked at your patch.  But I can give you an insight as to why the code is the way it is.
The parser is doing classic parse-matching: constantly rewriting the left end of the pattern, and consuming inputs and matching against symbols in those patterns.   I *think* of it as the left end of the pattern, but it can be flipped around so that input symbols match against the end of a vector instead of the start of a dequeue.  Perhaps that's exactly what you've done.  :-)

File: source/binary.cpp [hunk 3]
```diff
@@ -266,25 +274,15 @@ spv_result_t Parser::parseInstruction() {
 
   const uint32_t first_word = peek();
 
-  // TODO(dneto): If it's too expensive to construct the following "words"
```
Comment (dneto0) [inline source/binary.cpp]: W00t

File: source/binary.cpp [hunk 4]
```diff
@@ -310,13 +308,13 @@ spv_result_t Parser::parseInstruction() {
   // has its own logical operands (such as the LocalSize operand for
   // ExecutionMode), or for extended instructions that may have their
   // own operands depending on the selected extended instruction.
-  spv_operand_pattern_t expected_operands(
-      opcode_desc->operandTypes,
-      opcode_desc->operandTypes + opcode_desc->numTypes);
+  _.expected_operands.clear();
+  for (auto i = 0; i < opcode_desc->numTypes; i++)
```
Comment (dneto0) [inline source/binary.cpp]: I expect i is int and numTypes field is uint16_t and you'll get a compile error.  Recommend doing:

```
auto types_array = &(opcode_desc->operandTypes[0]);
for (unsigned i =0, numTypes = opcode_desc->numTypes ; i < numTypes ; i++)  {
 ... =  types_array[ numTypes - i - 1]
```

File: source/operand.cpp (hunk 2 above)
Comment (dneto0) [inline source/operand.cpp]: Indeed it's a true corner case for making extra-weird test cases.  Don't bother making this fast.

**Comment by chrisforbes** [comment]
@dneto0 is indeed what this does :).

File: source/binary.cpp (hunk 4 above)
Comment (atgoo) [inline source/binary.cpp]: The compiler likely compares i not to numTypes, but to int(numTypes), because it doesn't know how to compare int and uint16_t. So it would compile and work correctly.

**Comment by chrisforbes** [comment]
Updated:
- Fix tests that work directly with patterns, to think about them as the other way around.
- Rename 'prepend' to 'push' everywhere. I think this is reasonably clear?
- Fix text parser (was building its patterns backward, which went poorly)
- Use spv_operand_pattern_t name consistently -- we had a vector in one place before.

File: test/operand_pattern_test.cpp [hunk 5]
```diff
@@ -199,20 +199,29 @@ TEST(AlternatePatternFollowingImmediate, MultipleNonResultIds) {
 TEST(AlternatePatternFollowingImmediate, ResultIdFront) {
   EXPECT_THAT(spvAlternatePatternFollowingImmediate(
                   {SPV_OPERAND_TYPE_RESULT_ID, SPV_OPERAND_TYPE_ID}),
-              Eq(spv_operand_pattern_t{SPV_OPERAND_TYPE_RESULT_ID,
+              Eq(spv_operand_pattern_t{SPV_OPERAND_TYPE_OPTIONAL_CIV,
```
Comment (atgoo) [inline test/operand_pattern_test.cpp]: Are the input parameters not supposed to be in the reversed order?
```
EXPECT_THAT(spvAlternatePatternFollowingImmediate(
    {SPV_OPERAND_TYPE_ID, SPV_OPERAND_TYPE_RESULT_ID}),
```
I don't know if it's important, but it looks like you are changing the original tests if the input parameters are given in the same order as before, but the function expects them to be reversed.

Comment (chrisforbes) [inline test/operand_pattern_test.cpp]: These tests are patterns in, patterns out. I chose to fix up just the results rather than reverse both input and output patterns. I believe we still have good coverage of all the places a result id may be, across this test and its two siblings, but exactly what behavior is covered by each test moves around a little. Is that OK?

Comment (atgoo) [inline test/operand_pattern_test.cpp]: It's ok with me. It just looked confusing.

File: source/operand.h [hunk 6]
```diff
@@ -27,7 +27,7 @@
 //
 // As we parse an instruction in text or binary form from left to right,
 // we pull and push from the front of the pattern.
```
Comment (dneto0) [inline source/operand.h]: "we pop and push at the end of the pattern vector.   Symbols later in the pattern vector are matched against the input before symbols earlier in the pattern vector are matched."

Probably also worth saying why:
"Using a vector in this way reduces memory traffic, which is good for performance."

File: source/operand.h [hunk 7]
```diff
@@ -65,8 +65,8 @@ bool spvOperandIsVariable(spv_operand_type_t type);
 // Inserts a list of operand types into the front of the given pattern.
```
Comment (dneto0) [inline source/operand.h]: "Append a list of operand types to the end of the pattern vector.  "

File: source/operand.cpp [hunk 8]
```diff
@@ -218,26 +219,28 @@ const char* spvOperandTypeStr(spv_operand_type_t type) {
...
 
-void spvPrependOperandTypesForMask(const spv_operand_table operandTable,
-                                   const spv_operand_type_t type,
-                                   const uint32_t mask,
-                                   spv_operand_pattern_t* pattern) {
+void spvPushOperandTypesForMask(const spv_operand_table operandTable,
+                                const spv_operand_type_t type,
+                                const uint32_t mask,
+                                spv_operand_pattern_t* pattern) {
   // Scan from highest bits to lowest bits because we will prepend in LIFO
```
Comment (dneto0) [inline source/operand.cpp]: prepend -> append

File: test/operand_pattern_test.cpp (hunk 5 above)
Comment (dneto0) [inline test/operand_pattern_test.cpp]: Confusing at first, but this works for me.

**Comment by chrisforbes** [comment]
Comments all updated as requested.

**Comment by chrisforbes** [comment]
Adjusted again to fix new mark-v code.

File: source/assembly_grammar.h [hunk 9]
```diff
@@ -95,17 +95,19 @@ class AssemblyGrammar {
...
 
-  // Inserts the operands expected after the given typed mask onto the front
+  // Inserts the operands expected after the given typed mask onto the end
   // of the given pattern.
   //
-  // Each set bit in the mask represents zero or more operand types that should
-  // be prepended onto the pattern. Operands for a less significant bit always
-  // appear before operands for a more significant bit.
+  // Each set bit in the mask represents zero or more operand types that
+  // should be prepended onto the pattern. Operands for a less significant
```
Comment (dneto0) [inline source/assembly_grammar.h]: prepended -> appended

**Comment by chrisforbes** [comment]
Fixed :)

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as 78338d5ba
